web3/utils/events.py

```diff
@@ -238,3 +253,168 @@ def remove_trailing_from_seq(seq, remove_value=None):
 normalize_topic_list = compose(
     remove_trailing_from_seq(remove_value=None),
     pop_singlets,)
+
+
+def is_indexed(arg):
+    if isinstance(arg, TopicArgumentFilter) is True:
+        return True
+    return False
+
+
+not_indexed = complement(is_indexed)
+
+
+class EventFilterBuilder:
+    formatter = None
+    _fromBlock = None
+    _toBlock = None
+    _address = None
+
+    def __init__(self, event_abi, formatter=None):
+        self.event_abi = event_abi
+        self.formatter = formatter
+        self.event_topic = self._initial_event_topic()
+        self.args = AttributeDict(_build_argument_filters_from_event_abi(event_abi))
+        self._ordered_arg_names = tuple(arg['name'] for arg in event_abi['inputs'])
+
+    @property
+    def fromBlock(self):
+        return self._fromBlock
+
+    @fromBlock.setter
+    def fromBlock(self, value):
+        self._fromBlock = value
+
+    @property
+    def toBlock(self):
+        return self._toBlock
+
+    @toBlock.setter
+    def toBlock(self, value):
+        self._toBlock = value
+
+    @property
+    def address(self):
+        return self._address
+
+    @address.setter
+    def address(self, value):
+        self._address = value
+
+    @property
+    @to_tuple
+    def indexed_args(self):
+        return filter(
+            is_indexed,
+            map(lambda x: self.args[x], self._ordered_arg_names))
+
+    @property
+    @to_tuple
+    def data_args(self):
+        return filter(
+            not_indexed,
+            map(lambda x: self.args[x], self._ordered_arg_names))
+
+    @property
+    def topics(self):
+        arg_topics = tuple(arg.match_values for arg in self.indexed_args)
+        return normalize_topic_list(cons(to_hex(self.event_topic), arg_topics))
+
+    @property
+    def data_argument_values(self):
+        if self.data_args is not None:
+            return tuple(arg.match_values for arg in self.data_args)
+        else:
+            return (None,)
+
+    @property
+    def filter_params(self):
+        params = {
+            "topics": self.topics,
+            "fromBlock": self.fromBlock,
+            "toBlock": self.toBlock,
+            "address": self.address
+        }
+        return valfilter(lambda x: x is not None, params)
+
+    def deploy(self, web3):
+        if web3 is None:
+            raise ValueError("Missing required argument 'web3'")
+
+        log_filter = web3.eth.filter(self.filter_params)
+        log_filter.filter_params = self.filter_params
+        log_filter.set_data_filters(self.data_argument_values)
+        log_filter.builder = self
+        if self.formatter is not None:
+            log_filter.log_entry_formatter = self.formatter
+        return log_filter
+
+    def _initial_event_topic(self):
```

Comment:
I'd take this off the class and just make it a stand-alone private module function and pass the `event_abi` in as an input.